kv: fix (the horribly broken) prev()

tbg opened the pull request:
prev() was broken in two ways:

- it had a one-off where `prev([a, b), b)` would  return `b` instead of `KeyMin`.
- in one path it was assembling the return from the start keys, not the end keys, which is
  wrong. When keys `>= c` have been handled, and your remaining batch is `[a, b)`, it would
  erroneously return `a` instead of `b`.

tschottdorf commented:
@nvanbenschoten please give this a super thorough review, and if you have the local-key archaeology handy, I'd appreciate pointers (I can do it myself but haven't so far).

bdarnell commented on pkg/kv/batch.go:
```diff
@@ -126,6 +126,11 @@ func truncate(ba roachpb.BatchRequest, rs roachpb.RSpan) (roachpb.BatchRequest,
 // affect keys larger than the given key.
+//
+// Informally, a call `prev(ba, k)` means: we've already executed the parts
+// of `ba` that intersect `[key, KeyMax)`; please tell me how far to the
```
s/key/k/

tschottdorf replied:
Done.

tschottdorf replied:
Done.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/18351)
<!-- Reviewable:end -->

bdarnell commented on pkg/kv/batch.go:
```diff
@@ -140,24 +145,38 @@ func prev(ba roachpb.BatchRequest, k roachpb.RKey) (roachpb.RKey, error) {
 		if err != nil {
 			return nil, err
 		}
+		// FIXINPR(tschottdorf,nvanbenschoten): yikes, this is totally paged out.
```
Range-local keys exist in the "gaps" between regular keys. `keys.Addr` finds the lower bound: the greatest regular key that is less than or equal to the input, and AddrUpperBound finds the smallest regular key that is greater than or equal to the input. 

Here, we're trying to expand `[StartKey, EndKey)` to find the keys that form the tightest (potential) range boundaries around the given keys. So we must use Addr on StartKey and AddrUpperBound on EndKey to avoid leaving EndKey outside the new bounds.

tschottdorf replied:
Thanks for the reminder, added misc commentary.

tschottdorf replied:
Done.

tschottdorf commented:
Ok, ready for final pass. summoning @nvanbenschoten

nvanbenschoten commented:
Reviewed 2 of 3 files at r2.
Review status: 2 of 4 files reviewed at latest revision, 11 unresolved discussions, all commit checks successful.

---

*[pkg/keys/keys.go, line 357 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/18351#-KtXfDB0iRC6vlWTRQaF:-KtXfDB1hp5emJXX_qPf:b42uvm9) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0860f35323c1a4dfa30e3cd6ee3cce8db039d62/pkg/keys/keys.go#L357)):*
> ```Go
> // Addr returns the address for the key, used to lookup the range containing
> // the key. In the normal case, this is simply the key's value. However, for
> // local keys, such as transaction records, range-spanning binary tree node
> ```

somewhat unrelated: the `range-spanning binary tree node pointers` is gone, right?

---

*[pkg/keys/keys.go, line 415 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/18351#-KtXfqkfz_F1yJo_Nl6u:-KtXfqkgCr3MnYIlYIu7:b-aog851) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0860f35323c1a4dfa30e3cd6ee3cce8db039d62/pkg/keys/keys.go#L415)):*
> ```Go
> // but is suitable for use as the EndKey of a span involving a range-local key.
> //
> // Logically, the keys are arranged as follows:
> ```

Mind adding a similar example to `Addr`?

---

*[pkg/kv/batch.go, line 128 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/18351#-KtXk3NNgLIcTp8U8acc:-KtXk3NNgLIcTp8U8acd:b-4lfr8x) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0860f35323c1a4dfa30e3cd6ee3cce8db039d62/pkg/kv/batch.go#L128)):*
> ```Go
> 
> // prev gives the right boundary of the union of all requests which don't
> // affect keys larger than the given key.
> ```

I think it would be valuable here and below to remind the reader that a right boundary is exclusive but a left boundary is inclusive. That's the subtlety that causes the logic in each to be different and what caused the one-off you fixed here.

---

*[pkg/kv/batch.go, line 130 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/18351#-KtXg_vN1CByFg25KTJV:-KtXg_vN1CByFg25KTJW:b-puwre5) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0860f35323c1a4dfa30e3cd6ee3cce8db039d62/pkg/kv/batch.go#L130)):*
> ```Go
> // affect keys larger than the given key.
> //
> // Informally, a call `prev(ba, k)` means: we've already executed the parts
> ```

Add something like this to `next` please.

---

*[pkg/kv/batch.go, line 153 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/18351#-KtXh2Y7Q0S_0fgNoelm:-KtXh2Y7Q0S_0fgNoeln:b-387uv8) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0860f35323c1a4dfa30e3cd6ee3cce8db039d62/pkg/kv/batch.go#L153)):*
> ```Go
> 		}
> 		if len(eAddr) == 0 {
> 			// This is unintuitive, but if we have a point request at `x=k` then we're done
> ```

`we're done with k` is vague. Let's expand this comment now before we all forget how anything in here works for another 18 months.

---

*[pkg/kv/batch.go, line 158 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/18351#-KtXj40EQ4M6VoKOEPaT:-KtXj40EQ4M6VoKOEPaU:b474qxy) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0860f35323c1a4dfa30e3cd6ee3cce8db039d62/pkg/kv/batch.go#L158)):*
> ```Go
> 			eAddr = addr
> 		}
> 		if !eAddr.Less(k) {
> ```

Examples for each branch here would make this a lot easier to skim. Your call.

---

*[pkg/kv/batch_test.go, line 45 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/18351#-KtXl46PHMInP9lbQjSh:-KtXl46PHMInP9lbQjSi:b-fotmvu) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0860f35323c1a4dfa30e3cd6ee3cce8db039d62/pkg/kv/batch_test.go#L45)):*
> ```Go
> 	max, min := string(roachpb.RKeyMax), string(roachpb.RKeyMin)
> 	abc := span("a", "", "b", "", "c", "")
> 	testCases := []struct {
> ```

None of these test cases test multiple eligible `candidates`, right?

---

*[pkg/kv/batch_test.go, line 136 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/18351#-KtXlwK7FHgveM8dMwJk:-KtXlwK7FHgveM8dMwJl:bn6rw4y) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0860f35323c1a4dfa30e3cd6ee3cce8db039d62/pkg/kv/batch_test.go#L136)):*
> ```Go
> 			expBW: min,
> 		},
> 		{spans: span(loc("b"), ""), key: "b\x00",
> ```

Is this case right? I would have expected `expBW=b\x00`. I must be mixed up somewhere.

I would think we'd want:

```
endKey := h.EndKey
if len(endKey) == 0 {
    endKey = h.Key
}
eAddr, err := keys.AddrUpperBound(endKey)
if err != nil {
    return nil, err
}
```

instead of

```
eAddr, err := keys.AddrUpperBound(h.EndKey)
if err != nil {
    return nil, err
}
if len(eAddr) == 0 {
    eAddr = addr
}
```

---

*[pkg/kv/batch_test.go, line 143 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/18351#-KtXlHBWEnPbyteQC2uS:-KtXlHBWEnPbyteQC2uT:batsvgs) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0860f35323c1a4dfa30e3cd6ee3cce8db039d62/pkg/kv/batch_test.go#L143)):*
> ```Go
> 
> 	for _, test := range testCases {
> 		t.Run("", func(t *testing.T) {
> ```

No subtest name?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/18351)*
<!-- Sent from Reviewable.io -->

tbg committed:
kv: fix (the horribly broken) prev()

prev() was broken in two ways:

- it had a one-off where `prev([a, b), b)` would  return `b` instead of `KeyMin`.
- in one path it was assembling the return from the start keys, not the end keys, which is
  wrong. When keys `>= c` have been handled, and your remaining batch is `[a, b)`, it would
  erroneously return `a` instead of `b`.

tschottdorf commented:
TFTR! PR is updated.

---

Review status: 0 of 4 files reviewed at latest revision, 11 unresolved discussions.

---

*[pkg/keys/keys.go, line 357 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/18351#-KtXfDB0iRC6vlWTRQaF:-KtY5BRIYz8fUvJV0TOH:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0860f35323c1a4dfa30e3cd6ee3cce8db039d62/pkg/keys/keys.go#L357)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

somewhat unrelated: the `range-spanning binary tree node pointers` is gone, right?
</blockquote></details>

Done.

---

*[pkg/keys/keys.go, line 415 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/18351#-KtXfqkfz_F1yJo_Nl6u:-KtY6gD_8oD3yt_eKVIe:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0860f35323c1a4dfa30e3cd6ee3cce8db039d62/pkg/keys/keys.go#L415)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Mind adding a similar example to `Addr`?
</blockquote></details>

Done.

---

*[pkg/kv/batch.go, line 128 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/18351#-KtXk3NNgLIcTp8U8acc:-KtY7HZvmpkulr78uvSC:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0860f35323c1a4dfa30e3cd6ee3cce8db039d62/pkg/kv/batch.go#L128)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

I think it would be valuable here and below to remind the reader that a right boundary is exclusive but a left boundary is inclusive. That's the subtlety that causes the logic in each to be different and what caused the one-off you fixed here.
</blockquote></details>

Done.

---

*[pkg/kv/batch.go, line 130 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/18351#-KtXg_vN1CByFg25KTJV:-KtY7I-eDDZ1ESDH3a9T:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0860f35323c1a4dfa30e3cd6ee3cce8db039d62/pkg/kv/batch.go#L130)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Add something like this to `next` please.
</blockquote></details>

Done.

---

*[pkg/kv/batch.go, line 153 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/18351#-KtXh2Y7Q0S_0fgNoelm:-KtY7dSOR5jd9GRNpAuD:b-haa1be) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0860f35323c1a4dfa30e3cd6ee3cce8db039d62/pkg/kv/batch.go#L153)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

`we're done with k` is vague. Let's expand this comment now before we all forget how anything in here works for another 18 months.
</blockquote></details>

But it's so much fun to find out again! Done.

---

*[pkg/kv/batch.go, line 158 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/18351#-KtXj40EQ4M6VoKOEPaT:-KtY95q6pP6goaE2jU3V:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0860f35323c1a4dfa30e3cd6ee3cce8db039d62/pkg/kv/batch.go#L158)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Examples for each branch here would make this a lot easier to skim. Your call.
</blockquote></details>

Done.

---

*[pkg/kv/batch_test.go, line 45 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/18351#-KtXl46PHMInP9lbQjSh:-KtYBjda-X_gmWGANqsM:bjfwunu) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0860f35323c1a4dfa30e3cd6ee3cce8db039d62/pkg/kv/batch_test.go#L45)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

None of these test cases test multiple eligible `candidates`, right?
</blockquote></details>

They do now.

---

*[pkg/kv/batch_test.go, line 136 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/18351#-KtXlwK7FHgveM8dMwJk:-KtY9gTsik23PL3wqcrv:bs8duvo) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0860f35323c1a4dfa30e3cd6ee3cce8db039d62/pkg/kv/batch_test.go#L136)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

Is this case right? I would have expected `expBW=b\x00`. I must be mixed up somewhere.

I would think we'd want:

```
endKey := h.EndKey
if len(endKey) == 0 {
    endKey = h.Key
}
eAddr, err := keys.AddrUpperBound(endKey)
if err != nil {
    return nil, err
}
```

instead of

```
eAddr, err := keys.AddrUpperBound(h.EndKey)
if err != nil {
    return nil, err
}
if len(eAddr) == 0 {
    eAddr = addr
}
```

</blockquote></details>

Yikes, you're right. Fixed and added a boatload more comments.

---

*[pkg/kv/batch_test.go, line 143 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/18351#-KtXlHBWEnPbyteQC2uS:-KtYAqm8Ujurrq_rfZK6:bqytoto) ([raw file](https://github.com/cockroachdb/cockroach/blob/e0860f35323c1a4dfa30e3cd6ee3cce8db039d62/pkg/kv/batch_test.go#L143)):*
<details><summary><i>Previously, nvanbenschoten (Nathan VanBenschoten) wrote…</i></summary><blockquote>

No subtest name?
</blockquote></details>

No, apparently this is a thing. The test harness will enumerate it like `#%d`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/18351#-:-KtYDJ1w0KKQ0Jm22vHY:bjf48fz)*
<!-- Sent from Reviewable.io -->

nvanbenschoten commented:
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> thanks for adding all the comments! Hopefully we won't have to touch this code again for a long long time. 

---

Reviewed 1 of 3 files at r1, 1 of 3 files at r2, 2 of 3 files at r3.
Review status: 3 of 4 files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/18351#-:-KtYGXTk31ChY4eTswjv:b-xv6b5g)*
<!-- Sent from Reviewable.io -->

knz commented:
This needs to go to 1.1, but perhaps 1.0 too (if we ever release 1.0.7)

tschottdorf commented:
@knz ack, l'll see to it.
